pg_replicate/src/pipeline/sinks/delta.rs

```diff
@@ -0,0 +1,143 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use tokio_postgres::types::PgLsn;
+use tracing::info;
+
+use super::{BatchSink, SinkError};
+use crate::{
+    clients::delta::DeltaClient,
+    conversions::{cdc_event::CdcEvent, table_row::TableRow},
+    pipeline::PipelineResumptionState,
+    table::{TableId, TableSchema},
+};
+use deltalake::arrow::error::ArrowError;
+use deltalake::{arrow::datatypes::Schema, DeltaTableError};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DeltaSinkError {
+    #[error("Arrow error: {0}")]
+    DeltaArrow(#[from] ArrowError),
+    #[error("delta error: {0}")]
+    Delta(#[from] DeltaTableError),
+    #[error("missing table schemas")]
+    MissingTableSchemas,
+    #[error("missing delta schemas")]
+    MissingDeltaSchemas(String),
+    #[error("missing table id: {0}")]
+    MissingTableId(TableId),
+    #[error("incorrect commit lsn: {0}(expected: {0})")]
+    IncorrectCommitLsn(PgLsn, PgLsn),
+    #[error("commit message without begin message")]
+    CommitWithoutBegin,
+}
+
+pub struct DeltaSink {
+    client: DeltaClient,
+}
+
+impl DeltaSink {
+    pub fn new(path: String) -> Self {
+        DeltaSink {
+            client: DeltaClient {
+                path,
+                table_schemas: None,
+                delta_schemas: None,
+            },
+        }
+    }
+}
+
+impl SinkError for DeltaSinkError {}
+
+#[async_trait]
+impl BatchSink for DeltaSink {
+    type Error = DeltaSinkError;
+    async fn get_resumption_state(&mut self) -> Result<PipelineResumptionState, Self::Error> {
+        Ok(PipelineResumptionState {
+            copied_tables: HashSet::new(),
+            last_lsn: PgLsn::from(0),
+        })
+    }
+
+    async fn write_table_schemas(
+        &mut self,
+        table_schemas: HashMap<TableId, TableSchema>,
+    ) -> Result<(), Self::Error> {
+        let mut delta_schema: HashMap<String, Arc<Schema>> = HashMap::new();
+
+        for (_, table_schema) in &table_schemas {
+            let table_name = DeltaClient::table_name_in_delta(&table_schema.table_name);
+
+            let schema = self
+                .client
+                .create_table_schema(table_name.clone(), &table_schema.column_schemas)
+                .await?;
+
+            delta_schema.insert(table_name.clone(), schema);
+        }
+        self.client.delta_schemas = Some(delta_schema);
+        self.client.table_schemas = Some(table_schemas);
+
+        Ok(())
+    }
+
+    async fn write_table_rows(
+        &mut self,
+        rows: Vec<TableRow>,
+        table_id: TableId,
+    ) -> Result<(), Self::Error> {
+        for row in rows {
+            self.client
+                .write_to_table(row, table_id, String::from("I"))
+                .await?;
+        }
```

Comment:
Ok for a toy impl but the idea of moving to the `BatchedSink` trait was performance. Think of a way to write a batch of events to the table instead of one at a time, otherwise it will be slow as molasses.